chore(docs): v4 migration page (VIV-1801) (#1776)
Vonage/vivid-3: docs/resources/v4-release-migration.md

> # vivid-4.0.0
> 
> V-4 is currently scheduled for the 1st July 2024.  
> Vivid v-4 is a small iteration, not a rewrite like the last major release.  
> This is a migration guide to make the upgrade easy :)

Reviewer: ```suggestion
# Vivid v4.0.0 Migration Guide

Release date: 1st July 2024.  
This document provides descriptions of the breaking changes in this release and instructions on how to prepare for them.
```

---

## Audio-player

> ### The change

Reviewer: Could we provide a short description of the change for each? 
For example: 

The audio-player will now stretch to fill it's container's width by default. Previously it was restricted to 350px.

If you do not wish to have gutters, set `gutters="none"` on tabs:

```js
<vwc-tabs gutters="none">...</vwc-tabs>
```

---

## Tooltip

### The change

|                           | V-3 default value | V-4 default value |
| ------------------------- | ----------------- | ----------------- |
| tooltip `max-inline-size` | `none`            | `30ch`            |

### How to get ready?

If you do not wish to have a set `max-inline-size` use the css-variable `--tooltip-max-inline-size`.

```js
vwc-tooltip {
	--tooltip-max-inline-size: initial;
}
```

---

## Vue Wrappers

### The change

Form elements will now forward their value props as DOM properties instead of attributes.

|                     | V-3                                                                       | V-4                                                            |
| ------------------- | ------------------------------------------------------------------------- | -------------------------------------------------------------- |
| Form elements props | the value prop holds the initial value and currentValue the current value | value holds the current value and initialValue the initial one |

### How to get ready?

> TBA: Look into how we help with an upgrade script?

Reviewer: Maybe mention something about the eslint-plugin? Ask Richard for some help with the text.

Author: I did, he said that he will send me a line or two for this section :)

- VTimePicker
- VFilePicker

`checked` and `currentChecked` props on:

- VCheckbox
- VSwitch
- VRadio

`currentStart` / `currentEnd` on:

- VDateRangePicker
- VRangeSlider

Usage of these props is discouraged. To set the current value, you should use the `v-model` instead:

```html
<VTextField v-model="value" />
<!-- Which is syntactic sugar for: -->
<VTextField
	:model-value="value"
	@update:model-value="$event => (value = $event)"
/>
```

Code using `v-model` or `modelValue` is not affected.

### The change

|                                                                   | V-3                                | V-4                    |
| ----------------------------------------------------------------- | ---------------------------------- | ---------------------- |
| `currentValue` / `currentChecked` / `currentStart` / `currentEnd` | Controls current value             | Removed                |
| `value`/`checked`                                                 | Inconsistent between Vue 2 / Vue 3 | Controls current value |
| `initialValue` / `defaultChecked`                                 | Doesn't exist                      | Controls initial value |

## How to get ready?

> - First, make sure you're on the latest version of Vivid 2.x.  

Reviewer: Should this say Vivid 2.x?

Author: suppose to be 3.x :)

# Vivid v4.0.0 Migration Guide

Release date: 1st July 2024.  
This document provides descriptions of the breaking changes in this release and instructions on how to prepare for them.

---

## Audio-player

### The change

The audio-player will now stretch to fill it's container's inline-size by default. Previously it was restricted to 350px.

|                                               | V-3      | V-4       |
| --------------------------------------------- | -------- | --------- |
| audio-player `max-inline-size`                | `350px`  | `100%`    |
| css-variable `--audio-player-min-inline-size` | `in-use` | `removed` |

### How to get ready?

set `max-inline-size` or `inline-size` directly on audio-player or on its parent.

---

## Data-grid

> Data-grid-cell will no longer have a fixed block-size but will adjust to its content and its text will not be trimmed by default ad before.

Reviewer 2: This is only for non-header cells:
```suggestion
Data-grid-cells outside of the table header will no longer have a fixed block-size but will adjust to its content and its text will not be trimmed by default as before.
```

|                           | V-3 default value | V-4 default value |
| ------------------------- | ----------------- | ----------------- |
| tooltip `max-inline-size` | `none`            | `30ch`            |

### How to get ready?

If you do not wish to have a set `max-inline-size` use the css-variable `--tooltip-max-inline-size`.

```js
vwc-tooltip {
	--tooltip-max-inline-size: initial;
}
```

---

## Vue Wrappers

Form elements like `VTextField` will now forward value props consistently as DOM properties.

This change will affect the props `value` and `currentValue` on:

- VTextField
- VTextArea
- VNumberField
- VSelect
- VCombobox
- VSlider
- VRangeSlider
- VDatePicker
- VDateRangePicker
- VTimePicker
- VFilePicker

`checked` and `currentChecked` props on:

- VCheckbox
- VSwitch
- VRadio

`currentStart` / `currentEnd` on:

- VDateRangePicker
- VRangeSlider

Usage of these props is discouraged. To set the current value, you should use the `v-model` instead:

```html
<VTextField v-model="value" />
<!-- Which is syntactic sugar for: -->
<VTextField
	:model-value="value"

> 	@update:model-value="$event => (value = $event)"

Reviewer 2: Actually needs to be this:
```suggestion
	@update:modelValue="$event => (value = $event)"
```

/>
```

Code using `v-model` or `modelValue` is not affected.

### The change

|                                                                   | V-3                                | V-4                    |
| ----------------------------------------------------------------- | ---------------------------------- | ---------------------- |
| `currentValue` / `currentChecked` / `currentStart` / `currentEnd` | Controls current value             | Removed                |
| `value`/`checked`                                                 | Inconsistent between Vue 2 / Vue 3 | Controls current value |
| `initialValue` / `defaultChecked`                                 | Doesn't exist                      | Controls initial value |

## How to get ready?

> <vwc-note connotation="information" icon="info-solid" headline="This is not available yet">
> It will be introduced in Version 4
> </vwc-note>

Reviewer 2: Not true, it will go in 3.x
```suggestion
```

- First, make sure you're on the latest version of Vivid 3.x.  
  Optionally, install the latest version of our new ESLint plugin, which can help with the migration.

> - Replace all usage of `currentValue` / `currentChecked` with `modelValue`.
> - Replace `currentStart` / `currentEnd` with `start` / `end`.  
>   The `no-current-value-attribute` rule of our ESLint plugin can perform this migration automatically.

Reviewer 2: Importantly, the rule applies to both points:
```suggestion
- Replace all usage of `currentValue` / `currentChecked` with `modelValue` and all usages `currentStart` / `currentEnd` with `start` / `end`.  
  The `no-current-value-attribute` rule of our ESLint plugin can perform this migration automatically.
```
